frontend/src/components/common/rich-text-editor/RichTextPasting.tsx

```diff
@@ -0,0 +1,487 @@
+        data = {
+          rows: rowCount,
+          cols: colCount,
+          row: i,
+          col: j,
+        }
+        found = true
+        break
+      }
+    }
+    if (found) break
+  }
+  return { type: 'table-cell', data }
+}
+
+/**
+ * Custom functions to converts html into draft.js data structure state
+ */
+const stateFromHtmlOptions: any = {
+  /**
+   * Converts custom block to draft data struct
+   * @param element html element
+   * @returns type of block and its metadata
+   */
+  customBlockFn: (element: any) => {
+    // filter for specific styles to store in data
+    const style = element.getAttribute('style') || ''
+    const styleDict = convertStyleStringToObject(style) || {}
+    const data = Object.fromEntries(
+      Object.entries(styleDict).filter(([k, _v]) => ['text-align'].includes(k))
+    ) as { [id: string]: any }
+
+    // empty elements get ignored, ensure there is a non-breaking space
+    if (['P'].includes(element.tagName)) {
+      if (element.textContent === '') {
+        element.innerHTML = '&nbsp;'
+      }
+    }
+
+    // list handling (word doc)
+    if (element.className.startsWith('MsoList')) {
+      return MsoListConversion(element, styleDict)
+    }
+
+    // list handling of li tags (google docs)
+    if (element.tagName === 'LI') {
+      return liTagListConversion(element)
+    }
+
+    //table handling
+    if (element.tagName === 'TD') {
+      return tableConversion(element)
+    }
+    return { data }
+  },
+}
+
+/**
+ * Change html tag of an element
+ * @param element html element
+ * @param tag desired html tag
+ * @returns state with list depth adjusted
+ */
+function changeTag(element: Element, tag: string) {
+  // prepare the elements
+  const newElem = document.createElement(tag)
+  const clone = element.cloneNode(true) as Element
+  // move the children from the clone to the new element
+  while (clone.firstChild) {
+    newElem.appendChild(clone.firstChild)
+  }
+  // copy the attributes
+  const attrs = clone.attributes
+  for (let i = attrs.length - 1; i >= 0; i--) {
+    newElem.setAttribute(attrs[i].name, attrs[i].value)
+  }
+  return newElem
+}
+
+/**
+ * modify DOM tree for table
+ * - inserts a new line before table if table is the first element
+ * - If a nested table has a single row and cell,
+ * switch it to a span as a single cell's contents of the outer table
+ * @param node table element
+ * @param isNestedBlock if block is nested
+ * @returns
+ */
+function modifyTableDOM(node: Element, isNestedBlock: boolean) {
+  if (!node.previousElementSibling) {
+    const newNode = document.createElement('p')
+    node.parentNode?.insertBefore(newNode, node)
+  }
+
+  // Render content of single cell nested table directly
+  if (isNestedBlock) {
+    const cells = node?.querySelectorAll('td')
+    const firstCellChild = cells[0].firstElementChild
+    if (
+      cells.length === 1 &&
+      firstCellChild &&
+      ['DVI', 'P'].includes(firstCellChild.tagName)
```

Comment:
Is `DVI` a typo of `div`? Cos I was looking it up and couldn't find anything about `dvi` tag in mso docs